File: solutions/python/space-age/1/space_age.py
from functools import partial

class SpaceAge:

    SECONDS_PER_YEAR = 31_557_600

    ORBITAL_PERIOD_FOR_PLANET = {
        'mercury': 0.2408467,
        'venus': 0.61519726,
        'earth': 1.0,
        'mars': 1.8808158,
        'jupiter': 11.862615,
        'saturn': 29.447498,
        'uranus': 84.016846,
        'neptune': 164.79132,
    }
    
    def __init__(self, seconds):
        self.seconds = seconds
        for planet in self.ORBITAL_PERIOD_FOR_PLANET:
            setattr(self, f'on_{planet}', partial(self.calculate_age, planet))

    def calculate_age(self, planet):
        period = self.ORBITAL_PERIOD_FOR_PLANET[planet]
        return round(self.seconds / (self.SECONDS_PER_YEAR * period), 2)        

File: solutions/prolog/space-age/1/space_age.pl
period("Mercury", 0.2408467).
period("Venus", 0.61519726).
period("Earth", 1.0).
period("Mars", 1.8808158).
period("Jupiter", 11.862615).
period("Saturn", 29.447498).
period("Uranus", 84.016846).
period("Neptune", 164.79132).

space_age(Planet, AgeSec, Years) :-
    period(Planet, Period),
    SecondsPerYear is 60 * 60 * 24 * 365.25,
    Years = AgeSec / (Period * SecondsPerYear).
